syntax = "proto3";

service Branch {
  rpc MsgDelivery (Request) returns (Response) {} //Endpoint to handle all customer/branch requests
  rpc GetEvents (Empty) returns (EventList) {} //Endpoint to get the list of logical clock events stored in the branch
}

message Request {
  Interface interface = 1; //The type of the request
  int32 id = 2; //Represents the Id of the Customer or Branch making the request
  int32 money = 3;
  int32 logical_clock = 4; //The logical clock value for the event
  int32 customer_request_id = 5;
}

enum Interface {
  Query = 0;
  Withdraw = 1;
  Deposit = 2;
  Propagate_Withdraw = 3;
  Propagate_Deposit = 4;
}

message Response {
  Interface interface = 1;
  int32 id = 2;
  int32 money = 3;
  ResponseStatus status = 4; // Success or Failure
}

enum ResponseStatus {
  Success = 0;
  Failure = 1;
}

message Empty {

}

message EventList {
  int32 id = 1;
  string type = 2;
  repeated Event events = 3;
}

message Event {
  int32 customer_request_id = 1;
  int32 logical_clock = 2;
  Interface interface = 3;
  string comment = 4;
}